### PR TITLE
perf(rust, python): Speed up `.is_in` and `in`

### PR DESCRIPTION
**Pull request by jonashaag**
This adds some special cases to `is_in` for speedup:

- For small lhs, just use a bunch of `==`. It's faster.
- For ordered rhs, use binary search if we estimate it to be faster than constructing the hash set. It uses a complexity theory based formula to choose the algorithm. I experimentally verified that this formula holds surprisingly accurately.
- For the hash set algorithm, keep the min and max of the values and use that to short circuit hash set lookup. This speeds up lookups when there is little range overlap and doesn't slow down the other cases. NB: This could even be added to `==` for ordered series where min/max are cheap.

### CODE REVIEW

**Comment by alexander-beedie** [comment]
Can you put together some benchmarks to demonstrate/validate the speedups and the choice of boundary conditions that enable them? 👍

File: crates/polars-core/Cargo.toml [hunk 1]
```diff
@@ -63,6 +63,7 @@ rolling_window = []
 rank = []
 diff = []
 pct_change = ["diff"]
+search_sorted = []
```
Comment (ritchie46) [inline crates/polars-core/Cargo.toml:66]: Search sorted should not be moved into `polars-core`. Rather `is_in` should be moved into `polars-ops`.

**Comment by jonashaag** [comment]
> Looks interesting... Can you put together some benchmarks to demonstrate/validate the speedups and the choice of boundary conditions that enable them? 👍

Happy to! I have some synthetic microbenchmarks used for dev but I wonder if you could suggest some maybe more realistic ones?

**Comment by jonashaag** [comment]
Some benchmarking results:

- I wasn't able to come up with a useful microbenchmark to prove this change, so I'll revert it:
  > For the hash set algorithm, keep the min and max of the values and use that to short circuit hash set lookup
- For the `equals()` change, see [this gist](https://gist.github.com/jonashaag/05d5ab56b268e406998682f23344192d). Results:

  | lhs size | geomean (lower is better) |
  |----------|---------------------------|
  | 5        | 0.57 |
  | 10       | 0.67 |
  | 15       | 0.81 |
  | 20       | 0.93 |
  | 25       | 1.02 |
  | 30       | 1.10 |
  | 50       | 1.49 |

  Notes: rhs size doesn't matter (I checked [1, 1e6])
- Binary search: TBD

**Comment by jonashaag** [comment]
Results for binary search ([this script](https://gist.github.com/jonashaag/4dba5ebf38c39c8a011026ecde2b903a)):

Screenshot 2023-08-01 at 16 51 59

I hid all rows between 0.85 and 1.15 because I consider them measurement artefacts.

You can see that whenever the formula says TRUE (= we should use bin search), actually using bin search is faster, and vice-versa. There is a tiny number of outliers.

The "New with limit vs old" column is measured with taking into account the formula, while the "New no limit vs old" always uses binary search.

**Comment by jonashaag** [comment]
I'll need some help with moving `is_in` to polars-ops. There seem to be various ways to implement a series operation:

- `pub fn approx_unique(s: &Series) -> PolarsResult<Series>` ([here](https://github.com/pola-rs/polars/blob/main/crates/polars-ops/src/series/ops/approx_unique.rs))
- ```rs
  pub trait ArgAgg {
    fn arg_min(&self) -> Option<usize>;
  }

  impl ArgAgg for Series {
  ```

  ([here](https://github.com/pola-rs/polars/blob/main/crates/polars-ops/src/series/ops/arg_min_max.rs))
- `pub fn is_unique(s: &Series) -> PolarsResult<BooleanChunked>` ([here](https://github.com/pola-rs/polars/blob/main/crates/polars-ops/src/series/ops/is_unique.rs))
- ```rs
  trait IsIn {
    fn is_in(&self, _other: &Series) -> PolarsResult<BooleanChunked>;
  }
  ...
  impl IsIn for StructChunked {
    fn is_in(&self, other: &Series) -> PolarsResult<BooleanChunked> {
  ```
  ([here](https://github.com/pola-rs/polars/blob/main/crates/polars-core/src/chunked_array/ops/is_in.rs))

Why do these variants exist? When would you take as input/out a Series/ChunkedArray? When would you make a free function and when would make a trait?

File: crates/polars-core/src/chunked_array/ops/is_in.rs [hunk 2]
```diff
@@ -1,44 +1,114 @@
...
+        ca.into_iter()
+            .map(|opt_val| opt_val.map_or(other_has_nulls, |ca_val| other.equal(ca_val).any()))
+            .collect_trusted()
+    } else if can_use_binary_search && ca.len() * (other.len().ilog2() as usize) < other.len() {
+        // For sorted `other`, binary search can be faster than constructing the set of values in `other`.
+        // Complexity of hash set: construction: O(other) * O(1) + lookups: O(ca) * O(1)
+        // Complexity of binary search: lookups: O(ca) * O(log2(other))
+        // O(hash set)       <  O(binary search)              <=>
+        // O(other) + O(ca)  <  O(ca) * O(log2(other))        <=>
+        // O(other)          <  O(ca) * (O(log2(other)) - 1)  <=>
```
Comment (0xbe7a) [inline crates/polars-core/src/chunked_array/ops/is_in.rs:39]: O(other) <  O(ca) * (O(log2(other)) - 1) <=>  O(other) <  O(ca) * O(log2(other))
 
 Is this a typo?

**Pull request closed without merging**